Fix script to run examples locally

cniethammer opened the pull request:
Fixes problems reported in #284 for the `examples/run-examples.sh` script and makes it work again.

cniethammer committed:
Fix script to run examples locally

Ignore comment lines and empty lines in the example lists file.

Signed-off-by: Christoph Niethammer <niethammer@hlrs.de>

HomesGH reviewed (commented):
Thank you for updating the script!
There is a minor mess with those test scripts as they are all kind of duplicated:

1. [run-validation.sh](https://github.com/ls1mardyn/ls1-mardyn/blob/master/checks/run-validation.sh)
2. [run-examples.sh](https://github.com/ls1mardyn/ls1-mardyn/blob/master/examples/run-examples.sh)
3. [workflow / GitHub Actions](https://github.com/ls1mardyn/ls1-mardyn/blob/98d77b7d31073a12f0166f6f7315767b5f294616/.github/workflows/ls1_test.yml#L165-L209)

The workflow file differs a bit from the first two scripts but does basically the same (besides also comparing the results with the master branch).
However we can think of removing the `run-validation.sh` script since `run-examples.sh` is more sophisticated.

cniethammer commented:
Yes, it would be good to unify the script to reduce code maintenance and use this then in the GitHub workflow as well. I will check and come back with an update for this PR.

cniethammer committed:
Improve run script for local example execution

- Added command line options
- Add one second timeout at end of test to easy interruption
- Updated default run parameters to match run-validation script

Signed-off-by: Christoph Niethammer <niethammer@hlrs.de>

cniethammer commented:
Improved the run-example script so it can be reused in the run-validation script.

The github workflow requires more logic than this simple local executor script as it checks for matching outputs from different branches. So I will not address this with this PR and I consider this to be ready for merging.

cniethammer committed:
Update examples/run-examples.sh

Co-authored-by: FG-TUM <FG-TUM@users.noreply.github.com>

cniethammer committed:
Use basename to determine program name

Co-authored-by: FG-TUM <FG-TUM@users.noreply.github.com>

cniethammer commented on examples/run-examples.sh:
```diff
@@ -1,36 +1,146 @@
-#mapfile -t all_examples < $EXAMPLE_LIST_FILE # requires bash4 ...
-declare -a all_examples
-while IFS= read -r; do all_examples+=("$REPLY"); done < $EXAMPLE_LIST_FILE
+TEMP=$(getopt -o vh --long "help,inputlist:,logfile:,mardyn_args:,mardyn_exe:,mpirun_args:,mpirun_exe:,verbose" -- $@)
```
Definitly not pretty but what the official example for getopt suggests...

HomesGH commented on checks/run-validation.sh:
```diff
@@ -1,31 +1,19 @@
-	fi
-	cd ${repoPath}/examples/$(dirname $i)
+cd ${repoPath}/examples
+./run-examples.sh -v --inputlist "${inputlist}" --logfile "${logfileName}"
```
You could also use the options `--{mardyn,mpirun}_{exe,args}` here. Or we could just get rid of the entire file since its use case is already covered by the default settings of `run-examples.sh`, isn't it?

cniethammer replied:
Good point. I could not find any use case inside the codebase/workflows so I would get rid of the validation.sh script then.

cniethammer replied:
Deleted validation.sh as suggested with my last commit 23242b8667a28020c5dbcb73b1b2d4f851d8c6ee

HomesGH commented on examples/run-examples.sh:
```diff
@@ -1,36 +1,148 @@
+MARDYN_ARGS="--steps 20 -v"
+MPIRUN_EXE="mpirun"
+MPIRUN_ARGS="-n 4 --oversubscribe"
 EXAMPLE_LIST_FILE=${EXAMPLE_LIST_FILE:=example-list.txt}
 LOGFILE=${LOGFILE:=$PWD/run-examples.log}
```
Why are you not using the following?
```suggestion
EXAMPLE_LIST_FILE="example-list.txt"
LOGFILE="$PWD/run-examples.log"
```

HomesGH commented on examples/run-examples.sh:
```diff
@@ -1,36 +1,148 @@
+       shift 2
+       continue
+       ;;
+    '--mardyn_args')
+       MARDYN_EXE="$2"
+       shift 2
+       continue
+       ;;
+    '--mpirun_exe')
+       MARDYN_EXE="$2"
+       shift 2
+       continue
+       ;;
+    '--mpirun_args')
+       MARDYN_EXE="$2"
+       shift 2
+       continue
```
Why does it store all the arguments in `MARDYN_EXE` ? This is imho wrong. Please test all options of this script before merging to master!
```suggestion
    '--mardyn_args')
       MARDYN_ARGS="$2"
       shift 2
       continue
       ;;
    '--mpirun_exe')
       MPIRUN_EXE="$2"
       shift 2
       continue
       ;;
    '--mpirun_args')
       MPIRUN_ARGS="$2"
       shift 2
       continue
```

cniethammer replied:
Dam ... will fix!

cniethammer committed:
Fix incorrect variable assignments

Co-authored-by: HomesGH <55833544+HomesGH@users.noreply.github.com>

cniethammer committed:
Comment default arguments that are specific to Open MPI

Co-authored-by: FG-TUM <FG-TUM@users.noreply.github.com>

cniethammer committed:
Fix parsing option values in run-examples script

Fix parameter expansion so option values are kept correctly.
This fixes issues handling multi-value option arguments, e.g., allowing now
> ./run-examples.sh --mpi_args "-n 2 --oversubscribe 2 ..."

Signed-off-by: Christoph Niethammer <niethammer@hlrs.de>

cniethammer committed:
Change command line argument names to use hyphens instead of underscores

Signed-off-by: Christoph Niethammer <niethammer@hlrs.de>

cniethammer committed:
Remove unused run-validation.sh in favour of the run-examples.sh

Signed-off-by: Christoph Niethammer <niethammer@hlrs.de>

cniethammer committed:
[examples] Fix missing -i option in run-exampels script

Signed-off-by: Christoph Niethammer <niethammer@hlrs.de>

cniethammer committed:
[examples] Add option to specify base examples directory in run-examples.sh

Signed-off-by: Christoph Niethammer <niethammer@hlrs.de>

FG-TUM reviewed (approved):
I still get notified when I'm mentioned 😉

cniethammer commented:
THX @FG-TUM - and good to know you are still around here :-)
